### PR TITLE
move BoundServiceAccountTokenVolume to beta

### PR DESCRIPTION
**Pull request by zshihang**
**What type of PR is this?**

/kind feature
/kind api-change

**What this PR does / why we need it**:

graduate feature BoundServiceAccountTokenVolume to beta as the criteria is met.

https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/1205-bound-service-account-tokens#alpha-beta
upgrade test is passing: https://k8s-testgrid.appspot.com/sig-auth-gce#upgrade-tests

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
The BoundServiceAccountTokenVolume feature has been promoted to beta, and enabled by default.
* This changes the tokens provided to containers at `/var/run/secrets/kubernetes.io/serviceaccount/token` to be time-limited, auto-refreshed, and invalidated when the containing pod is deleted.
* Clients should reload the token from disk periodically (once per minute is recommended) to ensure they continue to use a valid token. `k8s.io/client-go` version v11.0.0+ and v0.15.0+ reload tokens automatically.
* By default, injected tokens are given an extended lifetime so they remain valid even after a new refreshed token is provided. The metric `serviceaccount_stale_tokens_total` can be used to monitor for workloads that are depending on the extended lifetime and are continuing to use tokens even after a refreshed token is provided to the container. If that metric indicates no existing workloads are depending on extended lifetimes, injected token lifetime can be shortened to 1 hour by starting `kube-apiserver` with `--service-account-extend-token-expiration=false`.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/1205-bound-service-account-tokens

### CODE REVIEW

**Comment by liggitt** [comment]
the upgrade CI test results look good, thanks for seeing that through!

* bazel-test needs default RBAC roles regenerated for the now-enabled `system:controller:root-ca-cert-publisher`

* unit, integration, verify tests and kubeadm are not yet passing the required flags for service account token signing, so they are failing with the following error:

> service-account-issuer is a required flag when BoundServiceAccountTokenVolume is enabled, --service-account-signing-key-file and --service-account-issuer are required flags

**Comment by liggitt** [comment]
The release note needs to more explicitly indicate what new API server arguments are required (and that the BoundServiceAccountTokenVolume feature can be disabled if desired while in beta in lieu of providing the new arguments)

**Comment by zshihang** [comment]
> * unit, integration, verify tests and kubeadm are not yet passing the required flags for service account token signing, so they are failing with the following error:
> 
> > service-account-issuer is a required flag when BoundServiceAccountTokenVolume is enabled, --service-account-signing-key-file and --service-account-issuer are required flags

that's what i was confused: those flags should be validated if `TokenRequest` is enabled but looks like code only requires those flags when `BoundServiceAccountTokenVolume` is enabled. should we fix the validating logic?


> the BoundServiceAccountTokenVolume feature can be disabled if desired while in beta in lieu of providing the new arguments

the apiserver will fail to start if those arguments are missing. how this feature can be disabled by not providing the new arguments? or use `AllBeta=true,BoundServiceAccountTokenVolume=false` to disable?

**Comment by zshihang** [comment]
/retest

the two failures 
- https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/95667/pull-kubernetes-local-e2e/1319063923147673600/
- https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/95667/pull-kubernetes-bazel-build/1319063908043984896

now looks irrelevant to this PR.

/assign @liggitt

**Comment by zshihang** [comment]
pull-kubernetes-bazel-build failure: https://github.com/kubernetes/test-infra/issues/19662

**Comment by dims** [comment]
@liggitt does the release note look good to you? esp `To enable this feature` ... don't we enable by default for all things marked Beta?

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 1]
```diff
@@ -132,20 +132,22 @@ func CreateStaticPodFiles(manifestDir, patchesDir string, cfg *kubeadmapi.Cluste
...
-		"enable-bootstrap-token-auth":     "true",
-		"secure-port":                     fmt.Sprintf("%d", localAPIEndpoint.BindPort),
-		"allow-privileged":                "true",
-		"kubelet-preferred-address-types": "InternalIP,ExternalIP,Hostname",
+		"advertise-address":                localAPIEndpoint.AdvertiseAddress,
+		"insecure-port":                    "0",
+		"enable-admission-plugins":         "NodeRestriction",
+		"service-cluster-ip-range":         cfg.Networking.ServiceSubnet,
+		"service-account-key-file":         filepath.Join(cfg.CertificatesDir, kubeadmconstants.ServiceAccountPublicKeyName),
+		"service-account-issuer":           "https://kubernetes.default.svc",
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/controlplane/manifests.go:140]: isn't this defined in a constant in a repository somewhere?

Comment (liggitt) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:140]: the service account issuer is not expected to be a constant or to be the same for all clusters.

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 2]
```diff
@@ -197,6 +199,13 @@ func getAPIServerCommand(cfg *kubeadmapi.ClusterConfiguration, localAPIEndpoint
 		defaultArguments["feature-gates"] = fmt.Sprintf("%s=%t", features.IPv6DualStack, enabled)
 	}
 
+	// TODO: remove after BoundServiceAccountTokenVolume is GA.
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/controlplane/manifests.go:202]: isn't BoundServiceAccountTokenVolume going GA in 1.20?

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 3]
```diff
@@ -197,6 +199,13 @@ func getAPIServerCommand(cfg *kubeadmapi.ClusterConfiguration, localAPIEndpoint
...
+	if enabled, ok := cfg.FeatureGates["AllBeta"]; ok && !enabled || strings.Contains(cfg.APIServer.ExtraArgs["feature-gates"], "AllBeta=false") {
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/controlplane/manifests.go:203]: kubeadm does not use the same list of feature gates as other k8s components. it only has these:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/features/features.go#L30-L35

this branch should be removed, and users that wish to opt-out of BoundServiceAccountTokenVolume should manually delete the flags `service-account-issuer`, `service-account-signing-key-file` from the kube-apiserver manifests.

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 1 above)
Comment (neolit123) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:140]: kubeadm seems to be pinning this in the alt names of the serving certificate already:
https://github.com/kubernetes/kubernetes/blob/604569482fc34d010a33339f000e75ecbbf345c6/cmd/kubeadm/app/util/pkiutil/pki_helpers.go#L380

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 2 above)
Comment (liggitt) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:202]: no, beta... TokenRequestProjection is already beta which allows pods to make use of those volumes themselves. BoundServiceAccountTokenVolume is the feature that switches the default injected volume to be a bound token, and is going from alpha to beta in 1.20

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 3 above)
Comment (liggitt) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:203]: I agree this branch is not great ... this is probably added here because of the ga-only conformance job which sets AllBeta=false and expects clusters to start and pass conformance tests.

The following features are beta:
1. TokenRequest (enables the API server to serve the /tokens endpoint and accept the signing-key/issuer flags)
2. TokenRequestProjection (enables pods to request projected token volumes)
3. BoundServiceAccountTokenVolume (enables the API server to inject service account tokens as projected token volumes by default)

Right now, the ga-only conformance job disables all three features with AllBeta=false. That means the API server complains about the signing-key/issuer flags. @mikedanese Moving TokenRequest/TokenRequestProjection to GA now seems reasonable to me, and would allow deployers like kubeadm to start populating the signing key and issuer flags.

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 1 above)
Comment (neolit123) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:140]: if this flag is supposed to be constructed from the cluster name then the following field can be used:
`cfg.NodeRegistration.Name`
(cfg is passed to the manifest constructing function)

[the remark this answers is not in the document]
Comment (neolit123) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:140]: > I expected this to be the public address of the API server

if this should be the address, then we should set `service-account-issuer` to the same value in:
`"advertise-address":                localAPIEndpoint.AdvertiseAddress,`

would it be fine that separate instances of the kube-apiserver in an HA setup to have separate value for `service-account-issuer` (if this is based on the address)?

> I don't know how these default values propagate into something user-modifiable. cc @neolit123 for insight there

nowadays, kubeadm allows patching the manifest files before the kubelet starts on control-plane nodes (to create the static pods), so technically users can patch the manifests with the flag changes they want to. it's also possible to use "phases", i.e. create manifests, edit manifests with text editor, start the pods later.

[the remark this answers is not in the document]
Comment (liggitt) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:140]: > would it be fine that separate instances of the kube-apiserver in an HA setup to have separate value for `service-account-issuer` (if this is based on the address)?

The issuer should be coherent between all servers in the cluster (or all servers should set an `--api-audiences` flag which includes all the servers' issuer names)

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 3 above)
Comment (neolit123) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:203]: thanks for the details!

> Moving TokenRequest/TokenRequestProjection to GA now seems reasonable to me, and would allow deployers like kubeadm to start populating the signing key and issuer flags.

ok, if not the branch can be adapted to something like the following, to conform the ga-only jobs without patches:

```
// TODO: remove after BoundServiceAccountTokenVolume is GA.
// ------> please add link to tracking issue here <------
if strings.Contains(cfg.APIServer.ExtraArgs["feature-gates"], "AllBeta=false") {
	delete(defaultArguments, "service-account-issuer")
	delete(defaultArguments, "service-account-signing-key-file")
}
```

but i cannot seem to be able to find the tracking issue for:
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/auth/bound-service-account-tokens.md

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (neolit123) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:140]: hm, populating `--api-audiences` with addresses from subsequent api-servers is a problem. statically bootstrapping etcd server instances has a similar problem, yet the `--initial-cluster` of old members don't have to be updated with the addresses of new members that join later.

> The issuer should be coherent between all servers in the cluster

it would be best if we can use the same value of `--service-account-issuer` on all instances, for the default setup.

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 3 above)
Comment (mikedanese) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:203]: > @mikedanese Moving TokenRequest/TokenRequestProjection to GA now seems reasonable to me, and would allow deployers like kubeadm to start populating the signing key and issuer flags

This sounds good to me.

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zshihang) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:140]: issuer is used as `iss` claim in the token. it would be a URL. https://tools.ietf.org/html/rfc7519#section-4.1.1

> server names can change and people can make mistakes and want to fix them

users can specify `api-audiences` if they doing the audiences change.

also, will also allow users to pass in `service-account-issuer` and the default will still be `https://kubernetes.default.svc`.

**Comment by zshihang** [comment]
most flags related work is done in https://github.com/kubernetes/kubernetes/pull/95896 per https://github.com/kubernetes/kubernetes/pull/95667#discussion_r512372282. also, the conformance test for TokenRequest requires that to pass.

File: test/cmd/apply.sh [hunk 4]
```diff
@@ -310,8 +310,8 @@ __EOF__
   kubectl delete -f hack/testdata/multi-resource-1.yaml "${kube_flags[@]:?}"
 
   ## kubectl apply multiple resources with one failure during builder phase.
-  # Pre-Condition: No configmaps
-  kube::test::get_object_assert configmaps "{{range.items}}{{${id_field:?}}}:{{end}}" ''
+  # Pre-Condition: Only "kube-root-ca.crt" configmap exists.
+  kube::test::get_object_assert configmaps "{{range.items}}{{${id_field}}}:{{end}}" 'kube-root-ca.crt:'
```
Comment (liggitt) [inline test/cmd/apply.sh:314]: scope the precondition to the actual config map we're inspecting:

```
kube::test::get_object_assert 'configmaps --field-selector=metadata.name=foo' "{{range.items}}{{${id_field:?}}}:{{end}}" ''
```

File: test/cmd/create.sh [hunk 5]
```diff
@@ -134,8 +134,9 @@ run_kubectl_create_kustomization_directory_tests() {
   set -o errexit
 
   ## kubectl create -k <dir> for kustomization directory
-  # Pre-condition: no ConfigMap, Deployment, Service exist
-  kube::test::get_object_assert configmaps "{{range.items}}{{$id_field}}:{{end}}" ''
```
Comment (liggitt) [inline test/cmd/create.sh:138]: scope to the `test-the-map` configmap with a field-selector

File: test/cmd/delete.sh [hunk 6]
```diff
@@ -37,17 +37,17 @@ run_kubectl_delete_allnamespaces_tests() {
   kubectl delete configmap --dry-run=client -l deletetest=true --all-namespaces
   kubectl delete configmap --dry-run=server -l deletetest=true --all-namespaces
   kubectl config set-context "${CONTEXT}" --namespace="${ns_one}"
-  kube::test::get_object_assert configmap "{{range.items}}{{${id_field:?}}}:{{end}}" 'one:'
+  kube::test::get_object_assert configmap "{{range.items}}{{${id_field:?}}}:{{end}}" 'kube-root-ca.crt:one:'
```
Comment (liggitt) [inline test/cmd/delete.sh:40]: scope the assertion to configmaps labeled with deletetest instead of explicitly adding references to `kube-root-ca.crt`

```
kube::test::get_object_assert 'configmap -l deletetest' "{{range.items}}{{${id_field:?}}}:{{end}}" 'one:'
```

File: test/cmd/get.sh [hunk 7]
```diff
@@ -206,8 +206,9 @@ run_kubectl_get_tests() {
   kubectl delete pods redis-master valid-pod "${kube_flags[@]}"
 
   ### Test 'kubectl get -k <dir>' prints all the items built from a kustomization directory
-  # Pre-condition: no ConfigMap, Deployment, Service exist
-  kube::test::get_object_assert configmaps "{{range.items}}{{$id_field}}:{{end}}" ''
```
Comment (liggitt) [inline test/cmd/get.sh:210]: scope the assertion to the `test-the-map` configmap

**Review by liggitt — COMMENTED** [review]
rather than adding references to the root ca configmap to the cmd tests, scope the assertions to the configmaps we actually care about (via --field-selector=metadata.name=$name or a label selector)

File: test/cmd/get.sh (hunk 7 above)
Comment (zshihang) [reply test/cmd/get.sh:210]: done

File: test/cmd/delete.sh (hunk 6 above)
Comment (zshihang) [reply test/cmd/delete.sh]: done

File: test/cmd/create.sh (hunk 5 above)
Comment (zshihang) [reply test/cmd/create.sh:138]: done

File: test/cmd/apply.sh (hunk 4 above)
Comment (zshihang) [reply test/cmd/apply.sh]: done

File: pkg/features/kube_features.go [hunk 8]
```diff
@@ -690,7 +690,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	HyperVContainer:                {Default: false, PreRelease: featuregate.Deprecated},
 	TokenRequest:                   {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // remove in 1.21
 	TokenRequestProjection:         {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // remove in 1.21
-	BoundServiceAccountTokenVolume: {Default: false, PreRelease: featuregate.Alpha},
+	BoundServiceAccountTokenVolume: {Default: true, PreRelease: featuregate.Beta},
```
[the remark this answers is not in the document]
Comment (zshihang) [reply pkg/features/kube_features.go:693]: https://github.com/kubernetes/website/pull/24854

[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/features/kube_features.go:693]: that's good, thanks. I also meant in the BoundServiceAccountTokenVolume definition above :-)

**Comment by liggitt** [comment]
Before merging this, I think `--service-account-extend-token-expiration` should be changed to true by default, per discussion in https://kubernetes.slack.com/archives/C0EN96KUY/p1603990482095800

The release note and website doc PR should call out the metrics that can be monitored to ensure workloads are not depending on the extended tokens, and how cluster-admins can turn off that extended token expiration once verified.

**Comment by zshihang** [comment]
> --service-account-extend-token-expiration should be changed to true by default

done

> call out the metrics that can be monitored to ensure workloads are not depending on the extended tokens, and how cluster-admins can turn off that extended token expiration once verified.

done for the release note

**Comment by liggitt** [comment]
Changes look good. https://github.com/kubernetes/kubernetes/pull/95896#issuecomment-718797458 is the only remaining question I have:

@neolit123:

> so given we are adding the flags for TokenRequest in 1.20, does this mean that if BoundServiceAccountTokenVolumes goes beta in 1.20, ideally we should keep it disabled (on the kubeadm side) until 1.21?

Unless `BoundServiceAccountTokenVolumes=true` or `AllBeta=true` is explicitly requested by the user, should kubeadm set `BoundServiceAccountTokenVolumes=false` for one release to allow the kubeadm TokenRequest enablement in https://github.com/kubernetes/kubernetes/pull/93258 to propagate to multi-server clusters? That would allow upgraded multi-server kubeadm clusters to work seamlessly with kubeadm defaults without any user involvement, which would match my expectations.

@zshihang, when you spoke with the various cloud providers / deployers for https://github.com/kubernetes/kubernetes/pull/95896#issuecomment-718216690, did they indicate whether they had already enabled TokenRequest in 1.19 or earlier releases?

**Comment by neolit123** [comment]
> Unless BoundServiceAccountTokenVolumes=true or AllBeta=true is explicitly requested by the user, should kubeadm set BoundServiceAccountTokenVolumes=false for one release to allow the kubeadm TokenRequest enablement in #93258 to propagate to multi-server clusters? That would allow upgraded multi-server kubeadm clusters to work seamlessly with kubeadm defaults without any user involvement, which would match my expectations.

based on my understanding, +1.
we should include the change in this PR.

is the BoundServiceAccountTokenVolumes FG only relevant to the kube-apiserver?

if so, it should be done bellow this line:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/controlplane/manifests.go#L205

```
// TODO: remove in 1.21. used for smooth transition in 1.19->1.20 upgrades around the BoundServiceAccountTokenVolume feature:
// https://github.com/kubernetes/enhancements/issues/2126
featureGates := cfg.APIServer.ExtraArgs["feature-gates"]
if !stings.Contains(featureGates, "BoundServiceAccountTokenVolume=true") && stings.Contains(featureGates, "AllBeta=true") {
	cfg.APIServer.ExtraArgs["feature-gates"] = fmt.Sprintf("%s,%s", featureGates, "BoundServiceAccountTokenVolume=false")
}
```

if the FG also matters to other components i can share links to the code.

**Comment by liggitt** [comment]
> is the BoundServiceAccountTokenVolume FG only relevant to the kube-apiserver?

In kube-apiserver, it enables auto-injection of the bound token volumes for pods. This is what kubeadm would want to wait until 1.21 to enable.

In kube-controller-manager, it enables the controller that maintains the kube-apiserver CA bundle configmap. Thinking about this more, I actually think this should generally be enabled a release ahead of the admission injection, which implies it should be a distinct gate. Otherwise, an n-1 kube-controller-manager would not produce the configmaps required for pods with the new admission-injected volumes to start.

/hold until sequencing between enabling the CA controller and the admission-injection is resolved

**Comment by neolit123** [comment]
> In kube-apiserver, it enables auto-injection of the bound token volumes for pods. This is what kubeadm would want to wait until 1.21 to enable.

ok

> In kube-controller-manager, it enables the controller that maintains the kube-apiserver CA bundle configmap. Thinking about this more, I actually think this should generally be enabled a release ahead of the admission injection, which implies it should be a distinct gate. Otherwise, an n-1 kube-controller-manager would not produce the configmaps required for pods with the new admission-injected volumes to start.

ok, if BoundServiceAccountTokenVolume goes beta in 1.20 and kubeadm passes `BoundServiceAccountTokenVolume=false` to the server and `BoundServiceAccountTokenVolume=true` (default) to the KCM i guess, it can workaround this problem, since 1.20 will be the release where the CM(CA) managing controller will be enabled and 1.21 when the injections will start to happen.

but for non-kubeadm users that is a problem and it does seem like a distinct gate is needed.

**Comment by liggitt** [comment]
I agree kubeadm can do fancy footwork to roll it out, but I'd prefer no fancy footwork be required and it "just work" by default.

**Comment by zshihang** [comment]
oops, i missed the kubeadm patch. done now.

> did they indicate whether they had already enabled TokenRequest in 1.19 or earlier releases?

yes.

**Comment by neolit123** [comment]
>> when you spoke with the various cloud providers / deployers for #95896 (comment), did they indicate whether they had already enabled TokenRequest in 1.19 or earlier releases?

> yes.

pretty sure this excludes VMware :)

**Comment by zshihang** [comment]
> > > when you spoke with the various cloud providers / deployers for #95896 (comment), did they indicate whether they had already enabled TokenRequest in 1.19 or earlier releases?
> 
> > yes.
> 
> pretty sure this excludes VMware :)

hmm, i confirmed with @dims that change is ok for VMware on slack.

**Comment by liggitt** [comment]
> i confirmed with @dims that change is ok for VMware on slack

sure, but "ok" != "enabled in 1.19".

**Comment by zshihang** [comment]
ok, sweeping my slack history:

(TokenRequest enabled FG turned on and signer flag is passed to kube-apiserver) in 1.19 or before:

- [x] Google
- [x] AWS (confirmed with @micahhausler )
- [ ] Azure (looks like the flags are optional to users. https://github.com/Azure/aks-engine/pull/3989)
- [ ] VMware (confirned with @neolit123)
- [x] Red Hat (confirmed with @deads2k, they have already been using bound token feature. double checking with him)
- [x] IBM (confirmed with @rtheis )


also, summarize the next steps before merge this change in 1.21:

1. mark flags required? (as all cloud providers agree on this change and also gives us confidence in multi server upgrades that all 1.20 clusters have TokenRequest enabled before BoundServiceAccountTokenVolume beta in 1.21)
2. add RootCAConfigMap FG to make sure kube-controller-manager publishes root-ca configmap before BoundServiceAccountTokenVolume beta.

**Comment by dims** [comment]
@liggitt happy to go with whatever the sig decides (wearing VMware hat). cc @andrewsykim @neolit123

File: pkg/kubeapiserver/options/authentication.go [hunk 9]
```diff
@@ -165,7 +165,7 @@ func (o *BuiltInAuthenticationOptions) WithRequestHeader() *BuiltInAuthenticatio
 
 // WithServiceAccounts set default value for service account authentication
 func (o *BuiltInAuthenticationOptions) WithServiceAccounts() *BuiltInAuthenticationOptions {
-	o.ServiceAccounts = &ServiceAccountAuthenticationOptions{Lookup: true}
+	o.ServiceAccounts = &ServiceAccountAuthenticationOptions{Lookup: true, ExtendExpiration: true}
```
Comment (liggitt) [inline pkg/kubeapiserver/options/authentication.go:168]: move this to a separate PR for merging in 1.20

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 10]
```diff
@@ -202,6 +202,14 @@ func getAPIServerCommand(cfg *kubeadmapi.ClusterConfiguration, localAPIEndpoint
 	if cfg.APIServer.ExtraArgs == nil {
 		cfg.APIServer.ExtraArgs = map[string]string{}
 	}
+
+	// TODO: remove in 1.21. used for smooth transition in 1.19->1.20 upgrades around the BoundServiceAccountTokenVolume feature:
```
Comment (liggitt) [inline cmd/kubeadm/app/phases/controlplane/manifests.go:206]: BoundServiceAccountTokenVolume to beta will be held for 1.21, so this change can be dropped

**Review by liggitt — COMMENTED** [review]
/milestone v1.21

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 10 above)
Comment (zshihang) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: done

File: pkg/kubeapiserver/options/authentication.go (hunk 9 above)
Comment (zshihang) [reply pkg/kubeapiserver/options/authentication.go]: done

**Comment by liggitt** [comment]
lgtm, will hold for 1.21
